### PR TITLE
Feature/18298 pupil data service refactor 2

### PR DESCRIPTION
**Pull request by jon-shipley**
Data layer refactor: pupils not taking check

### CODE REVIEW

File: admin/controllers/school.js [hunk 1]
```diff
@@ -439,62 +400,39 @@ const savePupilNotTakingCheck = async (req, res, next) => {
...
-      _id: req.body.attendanceCode,
-      dateRecorded: new Date(todayDate),
-      byUserName: req.user.UserName,
-      byUserEmail: req.user.EmailAddress
+  // The req.body.pupil data is posted in 3 forms:
+  // 1: string: 'abc-def' (single selection)
+  // 2: array of strings: ['abc-def', 'foo-bar'] (multiple selection)
+  // 3: object with properties/values: { 0: 'abc-def, 1: 'foo-bar' } (using checkbox "Select all")
+  let postedPupilSlugs
+  if (typeof req.body.pupil === 'object') {
```
Comment (GeorgeChatzigiannis) [inline admin/controllers/school.js]: how come there's an inconsistency on the data type of pupil coming from the frontend>

File: admin/data/sql/migrations/028.do.index-pupil-attendance.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+-- a pupil can only have one entry in the pupilAttendance table
+CREATE UNIQUE INDEX pupilAttendance_pupil_id_uindex ON mtc.mtc_admin.pupilAttendance (pupil_id);
```
Comment (GeorgeChatzigiannis) [inline admin/data/sql/migrations/028.do.index-pupil-attendance.sql]: might want to update your branch, this migration number is already used

File: admin/controllers/school.js (hunk 1 above)
Comment (jon-shipley) [reply admin/controllers/school.js]: String versus array depends on whether you pick one or more pupils.  I don’t know why it sometimes gets turned into an object.

File: admin/data/sql/migrations/028.do.index-pupil-attendance.sql (hunk 2 above)
Comment (jon-shipley) [reply admin/data/sql/migrations/028.do.index-pupil-attendance.sql]: Good spot!  Now renamed.

File: admin/controllers/school.js (hunk 1 above)
Comment (GeorgeChatzigiannis) [reply admin/controllers/school.js]: We'll definitely need to investigate this, it's very obnoxious.

File: admin/services/data-access/sql.service.js [hunk 3]
```diff
@@ -394,4 +399,19 @@ sqlService.update = async function (tableName, data) {
...
+ */
+sqlService.whereClauseHelper = (ary, type) => {
+  const params = []
+  const paramIdentifiers = []
+  for (let i = 0; i < ary.length; i++) {
+    params.push({ name: `p${i}`, type, value: ary[i] })
+    paramIdentifiers.push(`@p${i}`)
+  }
+  return {params, paramIdentifiers}
+}
```
Comment (GuyHarwood) [inline admin/services/data-access/sql.service.js]: nice 👌 
maybe consider naming it something more specific such as 'buildParameterList' or something similar?

**Pull request merged**